File: src/main/java/com/lalkalol/mujahid/util/Format.java
package com.lalkalol.mujahid.util;

import java.time.Duration;

public final class Format {
    private Format() {
    }

    public static String duration(long ms) {
        return duration(ms, false);
    }

    public static String duration(long ms, boolean isStream) {
        if (isStream) {
            return "LIVE";
        }
        Duration d = Duration.ofMillis(ms);
        long hours = d.toHours();
        int minutes = d.toMinutesPart();
        int seconds = d.toSecondsPart();
        if (hours > 0) {
            return String.format("%d:%02d:%02d", hours, minutes, seconds);
        }
        return String.format("%d:%02d", minutes, seconds);
    }

    public static String progressBar(long position, long total) {
        return progressBar(position, total, 18);
    }

    public static String progressBar(long position, long total, int length) {
        if (total <= 0) {
            return "🔘" + "▬".repeat(length - 1);
        }
        double ratio = Math.clamp(position / (double) total, 0.0, 1.0);
        int pos = (int) (ratio * (length - 1));
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < length; i++) {
            sb.append(i == pos ? "🔘" : "▬");
        }
        return sb.toString();
    }

    public static Long parseTimestamp(String input) {
        String[] parts = input.trim().split(":");
        for (String part : parts) {
            if (part.isBlank()) {
                return null;
            }
            try {
                Long.parseLong(part);
            } catch (NumberFormatException e) {
                return null;
            }
        }

        long seconds;
        switch (parts.length) {
            case 1 -> seconds = Long.parseLong(parts[0]);
            case 2 -> seconds = Long.parseLong(parts[0]) * 60 + Long.parseLong(parts[1]);
            case 3 -> seconds = Long.parseLong(parts[0]) * 3600
                    + Long.parseLong(parts[1]) * 60
                    + Long.parseLong(parts[2]);
            default -> {
                return null;
            }
        }
        return seconds * 1000;
    }
}

File: src/main/kotlin/com/lalkalol/mujahid/util/Format.kt
package com.lalkalol.mujahid.util

import java.time.Duration

object Format {
    fun duration(ms: Long): String = duration(ms, false)

    fun duration(ms: Long, isStream: Boolean): String {
        if (isStream) {
            return "LIVE"
        }
        val d = Duration.ofMillis(ms)
        val hours = d.toHours()
        val minutes = d.toMinutesPart()
        val seconds = d.toSecondsPart()
        return if (hours > 0) {
            String.format("%d:%02d:%02d", hours, minutes, seconds)
        } else {
            String.format("%d:%02d", minutes, seconds)
        }
    }

    fun progressBar(position: Long, total: Long): String = progressBar(position, total, 18)

    fun progressBar(position: Long, total: Long, length: Int): String {
        if (total <= 0) {
            return "🔘" + "▬".repeat(length - 1)
        }
        val ratio = (position.toDouble() / total).coerceIn(0.0, 1.0)
        val pos = (ratio * (length - 1)).toInt()
        return buildString {
            for (i in 0 until length) {
                append(if (i == pos) "🔘" else "▬")
            }
        }
    }

    fun parseTimestamp(input: String): Long? {
        val parts = input.trim().split(":")
        for (part in parts) {
            if (part.isBlank()) {
                return null
            }
            part.toLongOrNull() ?: return null
        }

        val seconds = when (parts.size) {
            1 -> parts[0].toLong()
            2 -> parts[0].toLong() * 60 + parts[1].toLong()
            3 -> parts[0].toLong() * 3600 + parts[1].toLong() * 60 + parts[2].toLong()
            else -> return null
        }
        return seconds * 1000
    }
}
